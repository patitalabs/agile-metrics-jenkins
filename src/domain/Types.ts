import {CiCollectorConfig, CiMetricItem} from "./ci/Types";

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface CiCollectorService {
  fetch(ciCollectorConfig: CiCollectorConfig): Promise<CiMetricItem[]>;
}
