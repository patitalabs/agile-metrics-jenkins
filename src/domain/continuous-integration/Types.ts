
export interface CiMetricItem {
  jenkinsUrl: string;
  buildName: string;
  result: string;
  durationInSeconds: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  projectName: string;
}

export class CiCollectorConfig  {
  orgName: string;
  projectName: string;
  teamName: string;
  since: Date;
  until?: Date;

  constructor({ orgName, projectName, since, until = null, teamName }) {
    this.orgName = orgName;
    this.projectName = projectName;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.teamName = teamName;
  }
}