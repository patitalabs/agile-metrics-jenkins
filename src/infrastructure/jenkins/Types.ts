export class JenkinsJob {
  name: string;
  branch: string;
  builds: JenkinsBuild[];
}

export class JenkinsBuild {
  timestamp: Date;
  result: string;
  durationInMs: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  url: string;
}

export interface JenkinsService {
  findData(
    orgName: string,
    projectName: string,
    since: Date,
    until?: Date
  ): Promise<JenkinsJob>;
}

export interface JenkinsClient {
  getData(url): Promise<any>;
}

export interface JenkinsRepository {
  jobDetails(
    orgName: string,
    projectName,
    since: Date,
    until?: Date
  ): Promise<JenkinsJob>;
}
