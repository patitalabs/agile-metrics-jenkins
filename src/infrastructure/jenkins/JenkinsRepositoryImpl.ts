import {
  JenkinsBuild,
  JenkinsClient,
  JenkinsJob,
  JenkinsRepository,
} from './Types';
import { Converters } from './Converters';
import { Utils } from '../../utils/Utils';

export class JenkinsRepositoryImpl implements JenkinsRepository {
  constructor(private readonly jenkinsClient: JenkinsClient) {}

  async jobDetails(
    orgName: string,
    projectName,
    since: Date,
    until?: Date
  ): Promise<JenkinsJob> {
    const jobDetails = await this.jenkinsClient.getData(
      `job/${orgName}/job/${projectName}/job/master`
    );

    const buildsPromises: Promise<JenkinsBuild>[] = jobDetails.builds.map(
      (jobBuild) => this.buildDetails(orgName, projectName, jobBuild.number)
    );

    const builds = await Promise.all(buildsPromises);
    const filteredBuilds = builds.filter((jobBuild) =>
      Utils.isDateInRange({
        createdAt: jobBuild.timestamp,
        since,
        until,
      })
    );
    return Converters.toJenkinsJob(jobDetails, filteredBuilds);
  }

  private async buildDetails(
    orgName: string,
    projectName: string,
    buildNumber: number
  ): Promise<JenkinsBuild> {
    const jobDetails = await this.jenkinsClient.getData(
      `job/${orgName}/job/${projectName}/job/master/${buildNumber}`
    );

    return Converters.toJenkinsBuild(jobDetails);
  }
}
