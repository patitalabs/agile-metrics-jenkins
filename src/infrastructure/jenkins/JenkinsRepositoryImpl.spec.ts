import * as projectBuildDetailsResponse from './test/project-build-details-response.json';
import * as projectDetailsResponse from './test/project-details-response.json';
import * as projectListResponse from './test/project-list-response.json';
import { JenkinsClient, JenkinsRepository } from './Types';
import { JenkinsRepositoryImpl } from './JenkinsRepositoryImpl';

describe('JenkinsRepository', () => {
  const jenkinsClient: JenkinsClient = {
    getData: (url): Promise<any> => {
      let result: any = projectListResponse;
      if (/.*job\/master\/\d+/.test(url)) {
        result = projectBuildDetailsResponse;
      } else if (/.*job\/master/.test(url)) {
        result = projectDetailsResponse;
      }
      return Promise.resolve(result);
    },
  };
  const jenkinsRepository: JenkinsRepository = new JenkinsRepositoryImpl(
    jenkinsClient
  );

  it('should get jobDetails data', async () => {
    const jenkinsJobs = await jenkinsRepository.jobDetails(
      'orgName',
      'project-name',
      new Date('2018-11-27')
    );

    expect(jenkinsJobs).toMatchSnapshot();
  });
});
