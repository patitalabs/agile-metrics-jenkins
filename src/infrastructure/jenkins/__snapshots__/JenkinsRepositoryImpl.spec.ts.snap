// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JenkinsRepository should get jobDetails data 1`] = `
Object {
  "branch": "master",
  "builds": Array [
    Object {
      "buildNumber": 44,
      "causedBy": "someuser",
      "durationInMs": 464989,
      "result": "SUCCESS",
      "revision": "9cc84c49ca9226620e35f5e850ca6e93568094fc",
      "revisionDescription": "org-6313 some comment",
      "timestamp": 2018-12-13T12:16:58.272Z,
      "url": "http://jenkins-server/job/org/job/project-name/job/master/44/",
    },
    Object {
      "buildNumber": 44,
      "causedBy": "someuser",
      "durationInMs": 464989,
      "result": "SUCCESS",
      "revision": "9cc84c49ca9226620e35f5e850ca6e93568094fc",
      "revisionDescription": "org-6313 some comment",
      "timestamp": 2018-12-13T12:16:58.272Z,
      "url": "http://jenkins-server/job/org/job/project-name/job/master/44/",
    },
    Object {
      "buildNumber": 44,
      "causedBy": "someuser",
      "durationInMs": 464989,
      "result": "SUCCESS",
      "revision": "9cc84c49ca9226620e35f5e850ca6e93568094fc",
      "revisionDescription": "org-6313 some comment",
      "timestamp": 2018-12-13T12:16:58.272Z,
      "url": "http://jenkins-server/job/org/job/project-name/job/master/44/",
    },
    Object {
      "buildNumber": 44,
      "causedBy": "someuser",
      "durationInMs": 464989,
      "result": "SUCCESS",
      "revision": "9cc84c49ca9226620e35f5e850ca6e93568094fc",
      "revisionDescription": "org-6313 some comment",
      "timestamp": 2018-12-13T12:16:58.272Z,
      "url": "http://jenkins-server/job/org/job/project-name/job/master/44/",
    },
    Object {
      "buildNumber": 44,
      "causedBy": "someuser",
      "durationInMs": 464989,
      "result": "SUCCESS",
      "revision": "9cc84c49ca9226620e35f5e850ca6e93568094fc",
      "revisionDescription": "org-6313 some comment",
      "timestamp": 2018-12-13T12:16:58.272Z,
      "url": "http://jenkins-server/job/org/job/project-name/job/master/44/",
    },
  ],
  "name": "org/project-name/master",
}
`;
